source/http/HttpConnectionManager.cpp

```diff
@@ -0,0 +1,294 @@
+                m_port = connectionManagerOptions.port;
+                m_socketOptions = *connectionManagerOptions.socketOptions;
+
+                if (connectionManagerOptions.tlsConnectionOptions)
+                {
+                    m_tlsConnOptions = *connectionManagerOptions.tlsConnectionOptions;
+
+                    if (!m_tlsConnOptions)
+                    {
+                        m_lastError = aws_last_error();
+                        m_good = false;
+                    }
+                }
+            }
+
+            HttpClientConnectionManager::~HttpClientConnectionManager()
+            {
+                Vector<std::shared_ptr<HttpClientConnection>> connectionsCopy = m_connections;
+                /* make sure all connections we know about are closed. */
+                for (auto &connection : connectionsCopy)
+                {
+                    connection->Close();
+                }
+
+                {
+                    /* wait for the last one to shutdown so the memory will be cleaned up. */
+                    std::unique_lock<std::mutex> m_semaphoreLock(m_connectionsLock);
+                    m_shutdownSemaphore.wait(m_semaphoreLock, [this] { return m_connections.empty(); });
+                }
+
+                /*
+                 * This would be really screwy if this ever happened. I don't even know what error to report, but in
+                 * case someone is waiting for a connection, AT LEAST let them know so they can not deadlock.
+                 */
+                for (auto &pendingConnectionRequest : m_pendingConnectionRequests)
+                {
+                    pendingConnectionRequest(nullptr, AWS_OP_ERR);
+                }
+            }
+
+            /* Asssumption: Whoever calls this already holds the lock. */
+            bool HttpClientConnectionManager::s_createConnection() noexcept
+            {
+                if (m_connections.size() + m_outstandingVendedConnections + m_pendingConnections < m_max_size)
+                {
+                    HttpClientConnectionOptions connectionOptions;
+                    connectionOptions.socketOptions = &m_socketOptions;
+                    connectionOptions.port = m_port;
+                    connectionOptions.hostName = ByteCursorFromCString(m_hostName.c_str());
+                    connectionOptions.bootstrap = m_bootstrap;
+                    connectionOptions.initialWindowSize = m_initialWindowSize;
+                    connectionOptions.allocator = m_allocator;
+                    connectionOptions.onConnectionSetup =
+                        [this](const std::shared_ptr<HttpClientConnection> &connection, int errorCode) {
+                            s_onConnectionSetup(connection, errorCode);
+                        };
+                    connectionOptions.onConnectionShutdown = [this](HttpClientConnection &connection, int errorCode) {
+                        s_onConnectionShutdown(connection, errorCode);
```

Comment:
here's a possible solution for the manager-must-outlive-its-connections problem:
```
std::weak_ptr<ConnectionManager> weakThis = weak_from_this();
connectionOptions.onConnectionShutdown = [weakThis](HttpClientConnection &connection, int errorCode) {
    if (auto sharedThis = weakThis.lock()) {
        sharedThis.s_onConnectionShutdown(connection, errorCode);
    }
}
```

The Manager would have to be shared_from_this, but I think this is pretty graceful.
This would remove the need for the semaphore stuff the Manager's destructor